File: 03.py
#!/usr/bin/env python3

from dataclasses import dataclass


@dataclass(frozen=True)
class Claim:
    id: int
    x: int
    y: int
    w: int
    h: int

    @classmethod
    def parse(cls, s):
        return cls(
            *map(int, s.translate(str.maketrans(',x', '  ', '#@:')).split()))

    def coords(self):
        for i in range(self.x, self.x + self.w):
            for j in range(self.y, self.y + self.h):
                yield (i, j)


def parse_claims(path):
    with open(path) as f:
        for line in f:
            yield Claim.parse(line)


def map_assignments(claims):
    ret = {}  # map coords to claims: (x, y) -> [claim]
    for claim in claims:
        for coord in claim.coords():
            ret.setdefault(coord, set()).add(claim)
    return ret


if __name__ == '__main__':
    claims = set(parse_claims('03.input'))
    assignments = map_assignments(claims)
    contested = {k: v for k, v in assignments.items() if len(v) > 1}

    # part 1
    print(len(contested))

    # part 2
    contested_claims = {c for claims in contested.values() for c in claims}
    uncontested_claims = claims - contested_claims
    assert len(uncontested_claims) == 1
    print(uncontested_claims.pop().id)

File: 03.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script

{-# OPTIONS_GHC -Wall #-}

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import qualified Data.IntSet as IntSet
import Text.ParserCombinators.ReadP
import Utils

data Claim = Claim {
    id_ :: Int,
    x :: Int,
    y :: Int,
    w :: Int,
    h :: Int }
    deriving Show

claim :: ReadP Claim
claim = do
    _id <- char '#' >> int
    _x <- string " @ " >> int
    _y <- char ',' >> int
    _w <- string ": " >> int
    _h <- char 'x' >> int <* eof
    return $ Claim _id _x _y _w _h

type Ids = [Int]

insertCoord :: Int -> Coord -> Map Coord Ids -> Map Coord Ids
insertCoord id' = Map.alter append where
    append Nothing = Just [id']
    append (Just v) = Just (id' : v)

insertClaim :: Claim -> Map Coord Ids -> Map Coord Ids
insertClaim claim_ m = foldr (insertCoord (id_ claim_)) m coords where
    coords = [(i, j) | i <- [start_x .. end_x], j <- [start_y .. end_y]]
    start_x = x claim_
    end_x = start_x + w claim_ - 1
    start_y = y claim_
    end_y = start_y + h claim_ - 1

assignments :: [Claim] -> Map Coord Ids
assignments = foldr insertClaim Map.empty

main :: IO ()
main = do
    input <- readFile "03.input"
    let claims = parseMany claim $ lines input -- mapMaybe parse (lines input)
    let assigned = assignments claims
    let contested = Map.filterWithKey (\_ v -> length v > 1) assigned
    -- part 1
    print $ Map.size contested
    -- part 2
    let allIds = IntSet.fromList [id_ c | c <- claims]
    let contestedIds = IntSet.fromList $ concatMap snd (Map.toList contested)
    print $ head $ IntSet.toList $ IntSet.difference allIds contestedIds
